### PR TITLE
Support Armv7

### PR DESCRIPTION
**Pull request by rchardx**
Support armv7 by adding armv7-compatible intrinsics.

* Compilation for Android: please refer to [armv7_compile.txt](https://github.com/ARM-software/astc-encoder/files/6115443/armv7_compile.txt).
  * Compiler:
> Android clang version 3.8.275480  (based on LLVM 3.8.275480)
> Target: x86_64-apple-darwin20.3.0
> Thread model: posix
> InstalledDir: /Users/.../Library/Android/sdk/ndk/14.1.3816874/toolchains/llvm/prebuilt/darwin-x86_64/bin/.
  * Option "-O3 -Wno-unused-command-line-argument" must be used for armv7.
* Unit Test: the results are shown in [unittest_armv7.txt](https://github.com/ARM-software/astc-encoder/files/6115456/unittest_armv7.txt)
  * Both aarch64 and armv7 fail vint4.two_to_the_n test.
  * Other fails are related to floating-point arithmetic.

### CODE REVIEW

File: Source/astcenc_vecmathlib_neon_armv7_4.h [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+	return vmulq_f32(a, x);
+}
+
+/**
+ * @brief Return a float rounded to the nearest integer value.
+ */
+ASTCENC_SIMD_INLINE float32x4_t vrndnq_f32(float32x4_t a)
+{
+	assert(::fegetround() == FE_TONEAREST);
+	float32_t a0 = ::nearbyintf(vgetq_lane_f32(a, 0));
```
Comment (solidpixel) [inline Source/astcenc_vecmathlib_neon_armv7_4.h:130]: Prefer these to have an explicit `std::nearbyint`.

File: Source/astcenc_vecmathlib_neon_armv7_4.h [hunk 2]
```diff
@@ -0,0 +1,174 @@
...
+	float32_t a1 = ::nearbyintf(vgetq_lane_f32(a, 1));
+	float32_t a2 = ::nearbyintf(vgetq_lane_f32(a, 2));
+	float32_t a3 = ::nearbyintf(vgetq_lane_f32(a, 3));
+	return {a0, a1, a2, a3};
```
Comment (solidpixel) [inline Source/astcenc_vecmathlib_neon_armv7_4.h:134]: Coding style nit - spaces between braces and fields please.

File: Source/astcenc_vecmathlib_neon_armv7_4.h [hunk 3]
```diff
@@ -0,0 +1,174 @@
...
+	int8x16_t id = vreinterpretq_s8_u8(idx);
+	return vcombine_s8(
+		vtbl2_s8(tab, vget_low_s8(id)),
+		vtbl2_s8(tab, vget_high_s8(id)));
+}
+
+/**
+ * @brief Add across Vector.
+ */
+ASTCENC_SIMD_INLINE uint32_t vaddvq_u32(uint32x4_t a)
```
Comment (solidpixel) [inline Source/astcenc_vecmathlib_neon_armv7_4.h:165]: Please remove vaddvq; it's no longer needed as I removed uses of it in the main aarch64 NEON code, and it's impossible to make invariant with x86 SIMD implementations.

File: Source/astcenc_vecmathlib_neon_armv7_4.h [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+	int32_t a3 = vgetq_lane_s32(a, 3);
+	return std::min(std::min(a0, a1), std::min(a2, a3));
+}
+
+/**
+ * @brief Return the sqrt of the lanes in the vector.
+ */
+ASTCENC_SIMD_INLINE float32x4_t vsqrtq_f32(float32x4_t a)
+{
+	a = vmaxq_f32(a, vdupq_n_f32((std::numeric_limits<float>::min)()));
```
Comment (solidpixel) [inline Source/astcenc_vecmathlib_neon_armv7_4.h:117]: Please replace with a scalarized version using `std::sqrt()`. Reciprocal sqrt is not invariant across implementations, and we have a strict invariance requirement.

**Review by solidpixel — CHANGES_REQUESTED** [review]
Thanks for the PR. Some code review comments added.

**Comment by solidpixel** [comment]
Note, I fixed the aarch64 NEON issue with the `vint4.two_to_the_n` unit test (invalid test - was shifting more than the data type could contain), so the unit tests should be clean.

**Review by solidpixel — COMMENTED** [review]
Do you have any build of this that works outside of a smartphone - e.g. for Arm Linux?

My main concern with accepting this as a PR is that I currently have no sensible way of testing it in our CI system. Our current aarch64 testing is on macOS, but I can't run 32-bit binaries there.

**Comment by rchardx** [comment]
I have fixed compilation issues on arm linux, including an assert error in two_to_the_n.
Current changes are ready to be compiled if vaddvq were removed.

* Compilation for Android Armv7
  * Script: [android_armv7_compile.txt](https://github.com/ARM-software/astc-encoder/files/6121456/android_armv7_compile.txt)
  * All tests passed!
* Compilation for ARM-based linux using arm-linux-gnueabi-g++
  * Env: Ubuntu 20.04, arm-linux-gnueabi-g++9.3
  * Script: [ubuntu_armv7_compile.sh.txt](https://github.com/ARM-software/astc-encoder/files/6121458/ubuntu_armv7_compile.sh.txt)
  * Executable file: [test-simd-neon_ubuntu2004_arm-linux-gnueabi-g++9.3.zip](https://github.com/ARM-software/astc-encoder/files/6121469/test-simd-neon_ubuntu2004_arm-linux-gnueabi-g%2B%2B9.3.zip)
  * Sadly, I have no ARM32 linux devices and therefore did not run the tests.

**Comment by solidpixel** [comment]
Thanks for the update. For running on Armv7 I've been using using `qemu-user` as an emulator. 

`
apt-get install qemu-user
export QEMU_LD_PREFIX=/usr/arm-linux-gnueabi
qemu-user ./astcenc/test-simd-neon
`

**Comment by solidpixel** [comment]
Still one test fail in debug builds (my fault). Let me merge that to `main` with the vaddvq removal change.

**Comment by rchardx** [comment]
> Thanks for the update. For running on Armv7 I've been using using `qemu-user` as an emulator on an x86-64 machine.
> 
> ```
> apt-get install qemu-user
> export QEMU_LD_PREFIX=/usr/arm-linux-gnueabi
> qemu-user ./astcenc/test-simd-neon
> ```

Armv7 on linux (release) passed all tests using qemu-arm.

**Review by solidpixel — APPROVED** [review]

**Comment by solidpixel** [comment]
Merged; still some work to do Arm-side to add some test automation using `qemu-user`, but I'll handle that as a separate commit.

**Comment by Lyt-BUPT** [comment]
why this support is delete in the new master?

**Comment by rchardx** [comment]
> Why this support is delete in the new master? Is there any new replace way to get .a library that can be used on android armv7 devices?

astc-encoder dropped the support of armv7 devices.
You can just use older versions of the repo.

**Comment by solidpixel** [comment]
The main problem is that it wasn't actually Armv7 complaint - it was really a hybrid of Armv7 and Armv8 AArch32. The main reason we removed it was that we had no way to test it, and increasingly 32-bit software is getting phased out so the cost of maintaining and adding testing for it doesn't make a huge amount of sense.

As above, old release tags are still live if needed.

**Comment by solidpixel** [comment]
https://github.com/ARM-software/astc-encoder/tree/3.0 is the last release tag that still supported this.

**Comment by Lyt-BUPT** [comment]
> > Why this support is delete in the new master? Is there any new replace way to get .a library that can be used on android armv7 devices?
> 
> astc-encoder dropped the support of armv7 devices. You can just use older versions of the repo.

I got it. Thanks for your replies!

**Comment by Lyt-BUPT** [comment]
> https://github.com/ARM-software/astc-encoder/tree/3.0 is the last release tag that still supported this.


I got it. Thanks for your replies!

**Pull request merged**